src/Two/AbstractProvider.php

```diff
@@ -200,9 +200,9 @@ protected function hasInvalidState()
             return false;
         }
 
-        $session = $this->request->getSession();
+        $state = $this->request->getSession()->pull('state');
 
-        return ! ($this->request->input('state') === $session->get('state'));
+        return ! (strlen($state) > 0 && $this->request->input('state') === $state);
```

Comment:
That's exactly what this PR fixes actually. "pull" from the session means we only read the value once, then it's gone, so someone can't try to guess it again.